### PR TITLE
sparse_mask: backward support for sparse lhs (take 2)

### PR DESCRIPTION
**Pull request by nikitaved**
This is a copy of https://github.com/pytorch/pytorch/pull/95165 with some bug fixes.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99298
* #99043
* #98490
* __->__ #104341

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

**Comment by nikitaved** [comment]
Sorry to bother you again with this one, @pearu, but could you please approve this one again?

**Comment by nikitaved** [comment]
@albanD , @soulitzer , could you please also approve it for good measure since @cpuhrsch is on PTO?

**Review by albanD — APPROVED** [review]
Looks good!

File: test/test_sparse.py [hunk 1]
```diff
@@ -2140,6 +2141,36 @@ def _test_sparse_mask_hybrid_fixed():
...
+                    lhs_data.shape
+                )._coalesced_(lhs_is_coalesced).requires_grad_(True)
+
+                rhs = torch.sparse_coo_tensor(
+                    lhs_data._indices()[:, -nnz:],
+                    lhs_data._values()[-nnz:],
+                    lhs_data.shape
+                )._coalesced_(rhs_is_coalesced)
+
+                # setting masked = True is required because of the broken backward of to_dense().
```
Comment (amjames) [inline test/test_sparse.py:2170]: Does this work as expected without using the `masked=True` kwarg for gradcheck, if you instead use the `masked_grad=False` kwarg to `to_dense` ?

Comment (nikitaved) [reply test/test_sparse.py:2170]: Yes, this change should be there up in the stack. But let me add it here for completeness.

File: test/test_sparse.py [hunk 2]
```diff
@@ -2140,6 +2141,36 @@ def _test_sparse_mask_hybrid_fixed():
...
+                # See https://github.com/pytorch/pytorch/issues/95550.
+                gradcheck(lambda x: x.sparse_mask(rhs).to_dense(), (lhs,), masked=True, check_sparse_nnz=True)
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply test/test_sparse.py]: Removed.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @nikitaved!

**Review by amjames — APPROVED** [review]
Thanks!

**Comment by nikitaved** [comment]
Alright, it looks weird. Slowtests are failing in somewhat non-deterministic order across the shards. Is there a way to trigger these things locally? I have no issues when running gradcheck with `fast_mode=False`.

**Comment by pearu** [comment]
> Alright, it looks weird. Slowtests are failing in somewhat non-deterministic manner across shards. Is there a way to trigger these things locally? I have no issues when running gradcheck with `fast_mode=False`.

Yes, try defining env variable `PYTORCH_TEST_WITH_SLOW_GRADCHECK=1` prior running the gradcheck tests.

**Comment by nikitaved** [comment]
Thank you, @pearu! Now it fails locally, but only when `masked=True`...

**Comment by nikitaved** [comment]
Nice, problem solved! One more for `masked=True` being very confusing ;)

**Comment by nikitaved** [comment]
The failures are unrelated :) I think it is good to go!

**Comment by nikitaved** [comment]
@amjames , I will rebase, then we can merge :)
